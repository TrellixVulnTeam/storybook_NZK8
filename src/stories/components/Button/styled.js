// @flow
/* eslint-disable import/prefer-default-export */

import styled from 'styled-components';
import { lighten } from 'polished';
import { colors } from 'utils/StyleGuide';
import { getGradientColors } from 'utils/StyleHelpers';

/**
 * Determine the button's accent color based on the "color" prop
 * @param {string} theme
 * @return {string}
 */
function getAccentColor(theme) {
  switch (theme) {
    case 'blue':
      return colors.brightBlue;
    case 'green':
      return colors.shamrockGreen;
    case 'dark-green':
      return colors.jungleGreen;
    case 'white':
      return colors.alabaster;
    case 'red':
      return colors.strawberry;
    case 'silver':
      return colors.silver;
    case 'cancel':
    case 'transparent':
    case 'filter':
    case 'filter-active':
    case 'blue-bordered':
    case 'steel':
    default:
      return colors.transparent;
  }
}

/**
 * Determine the button's font color based on the "theme" prop
 * @param {string} theme
 * @return {string}
 */
function getFontColor(theme) {
  switch (theme) {
    case 'transparent':
    case 'blue-bordered':
      return colors.brightBlue;
    case 'steel':
      return colors.steel;
    case 'red':
    case 'blue':
    case 'green':
    case 'blue-gradient':
    case 'red-gradient':
    case 'dark-green':
    case 'white-outline':
      return colors.white;
    case 'cancel':
    case 'white':
    case 'filter':
    case 'filter-active':
    default:
      return colors.slateGrey;
  }
}

/**
 * Determine the button's font color
 * on hover based on the "theme" prop
 * @param {string} theme
 * @return {string}
 */
function getHoverFontColor(theme) {
  switch (theme) {
    case 'red':
    case 'blue':
    case 'green':
    case 'dark-green':
    case 'blue-gradient':
    case 'red-gradient':
    case 'white-outline':
      return colors.white;
    case 'white':
    case 'cancel':
    case 'transparent':
    case 'filter':
    case 'filter-active':
    case 'blue-bordered':
    case 'steel':
    default:
      return colors.brightBlue;
  }
}

/**
 * Determine the button's border color based on the "theme" prop
 * @param {string} theme
 * @return {string}
 */
function getBorderColor(theme) {
  switch (theme) {
    case 'white':
      return '#ccc';
    case 'filter':
      return colors.silver;
    case 'filter-active':
    case 'blue-bordered':
      return colors.brightBlue;
    case 'white-outline':
      return colors.white;
    default:
      return '';
  }
}

/**
 * Determine the button's size based on the "size" prop
 * @param {string} size
 * @return {string}
 */
function getSize(size) {
  switch (size) {
    case 'xs':
      return { width: 36, height: 36, font: 12, vpad: 0, hpad: 0 };
    case 'sm':
      return { width: 60, height: 36, font: 12, vpad: 4, hpad: 8 };
    case 'lg':
      return { width: 70, height: 48, font: 16, vpad: 6, hpad: 20 };
    case 'xlg':
      return { width: 80, height: 42, font: 14, vpad: 8, hpad: 15 };
    case 'xl':
      return { width: 100, height: 60, font: 20, vpad: 8, hpad: 24 };
    case 'xxl':
      return { width: 150, height: 42, font: 14, vpad: 10, hpad: 42 };
    case 'xxxl':
      return { width: 360, height: 50, font: 16, vpad: 0, hpad: 0 };
    default:
      return { width: 100, height: 36, font: 14, vpad: 6, hpad: 16 };
  }
}

/**
 * Determine the button's box-shadow
 *  on hover based on the "theme" prop
 * @param {string} theme
 * @return {string} box-shadow property value
 */
function getBoxShadow(theme) {
  switch (theme) {
    case 'filter':
    case 'filter-active':
    case 'transparent':
      return '0 2px 10px 0 rgba(40, 44, 49, 0.2)';
    default:
      return 'none';
  }
}

/**
 * Determine the button's icon
 * color based on the "theme" prop
 * @param {string} theme
 * @return {string} fill property value
 */
function getFillColor(theme) {
  switch (theme) {
    case 'cancel':
    case 'blue-bordered':
      return colors.brightBlue;
    case 'red':
    case 'green':
    case 'dark-green':
    case 'blue':
    case 'red-gradient':
    case 'blue-gradient':
      return colors.white;
    case 'white':
    case 'transparent':
    case 'filter':
    case 'filter-active':
    case 'steel':
    default:
      return colors.steel;
  }
}

/**
 * Determine the button's icon
 * color based on the "theme" prop
 * @param {string} theme
 * @return {string} fill property value
 */
function getHoverFillColor(theme) {
  switch (theme) {
    case 'red':
    case 'green':
    case 'dark-green':
    case 'blue':
    case 'red-gradient':
    case 'blue-gradient':
      return colors.white;
    case 'white':
    case 'transparent':
    case 'filter':
    case 'filter-active':
    case 'cancel':
    case 'blue-bordered':
    case 'steel':
    default:
      return colors.brightBlue;
  }
}

export const Wrapper = styled.button`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  position: relative;
  margin: 0 6px;
  color: ${p => getFontColor(p.theme)};
  font-size: ${p => getSize(p.size).font}px;
  font-weight: ${p => (p.thin ? 400 : 600)};
  height: ${p => getSize(p.size).height}px;
  padding: ${p => getSize(p.size).vpad}px ${p => getSize(p.size).hpad}px;
  border-radius: ${p => (p.pill ? 30 : 6)}px;
  text-align: center;
  border: ${p =>
    getBorderColor(p.theme)
      ? `1px solid ${getBorderColor(p.theme)}`
      : `1px solid ${colors.transparent}`};
  background-clip: padding-box;
  background-color: ${({ theme, disabledBackground, disabled }) =>
    disabled && disabledBackground ? getAccentColor(disabledBackground) : getAccentColor(theme)};
  background-image: ${p =>
    getGradientColors(p.theme).length === 2
      ? `linear-gradient(251deg, ${getGradientColors(p.theme)[0]}, ${
          getGradientColors(p.theme)[1]
        })`
      : 'none'};
  opacity: ${p => (!p.loading && p.disabled ? 0.5 : 1)};
  letter-spacing: -0.34px;
  min-width: ${p => getSize(p.size).width}px;
  transition: all 150ms ease;

  svg {
    fill: ${p => getFillColor(p.theme)};
    margin-right: 5px;
  }

  &:hover {
    cursor: ${p => (p.disabled ? 'default' : 'pointer')};
    background-color: ${p =>
      p.theme === 'cancel' || p.theme === 'white-outline'
        ? 'rgba(255,255,255,0.12)'
        : lighten(0.15, getAccentColor(p.theme))};
    box-shadow: ${p => getBoxShadow(p.theme)};
    color: ${p => getHoverFontColor(p.theme)};
    background-image: ${p =>
      getGradientColors(p.theme).length === 2
        ? `linear-gradient(251deg, ${lighten(0.15, String(getGradientColors(p.theme)[0]))},
      ${lighten(0.15, String(getGradientColors(p.theme)[1]))})`
        : 'none'};
    border: ${p => p.theme === 'filter' && '1px solid transparent'};

    svg {
      fill: ${p => getHoverFillColor(p.theme)};
    }
  }

  &:focus {
    text-decoration: none;
    outline: none;
  }
`;
