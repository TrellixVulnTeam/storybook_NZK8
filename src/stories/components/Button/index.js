// @flow

import React from 'react';
import Loader from 'components/Loader';
import { Wrapper } from './styled';

// Themes
// 'white-outline'
// 'blue-gradient' | 'red-gradient'
// 'filter' | 'filter-active' | 'cancel'
// 'blue' | 'red' | 'green' | 'white' | 'transparent' | 'dark-green
type Props = {
  children: React$Node,
  onClick?: Function,
  theme?: string,
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xlg' | 'xl' | 'xxl' | 'xxxl',
  thin?: boolean,
  pill?: boolean,
  loading?: boolean,
  disabled?: boolean,
  disabledBackground?: string,
  type?: 'button' | 'submit' | 'reset',
  form?: string,
  className?: string,
};

export default (props: Props) => {
  return (
    <Wrapper
      theme={props.theme}
      size={props.size}
      thin={props.thin}
      pill={props.pill}
      onClick={(...args) => {
        if (props.disabled || props.loading || typeof props.onClick !== 'function') {
          return;
        }

        props.onClick(...args);
      }}
      disabled={props.loading || props.disabled}
      disabledBackground={props.disabledBackground}
      loading={props.loading}
      className={props.className}
      type={props.type || 'button'}
      form={props.form || undefined}
    >
      {props.loading ? <Loader size="sm" /> : props.children}
    </Wrapper>
  );
};
