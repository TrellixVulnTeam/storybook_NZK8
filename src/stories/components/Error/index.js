// @flow

import React, { PureComponent } from 'react';
import Button from 'components/Button';
import Page from 'components/Page';

import { PageTitle, Details } from './styled';

type Props = {
  title: string,
  error: string,
  errorInfo?: string,
  showReload?: boolean,
};

class Error extends PureComponent<Props> {
  /**
   * Render the Error component
   * @return {React$Node}
   */
  render() {
    return (
      <Page withMargin>
        <PageTitle>{this.props.title}</PageTitle>
        {this.props.showReload && (
          <Button theme="blue-gradient" pill onClick={() => window.location.reload()}>
            Refresh
          </Button>
        )}
        <Details>
          <summary>{this.props.error}</summary>
          <br />
          {this.props.errorInfo ? this.props.errorInfo : ''}
        </Details>
      </Page>
    );
  }
}

export default Error;
