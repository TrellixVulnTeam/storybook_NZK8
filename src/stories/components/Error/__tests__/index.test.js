import React from 'react';
import { mount } from 'enzyme';
import Error from 'components/Error/index';
import Button from 'components/Button';

describe('<Error />', () => {
  const mandatoryProps = {
    title: 'title',
    error: 'ERROR',
  };
  const reloadSpy = jest.spyOn(window.location, 'reload');

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with mandatory props', () => {
    const sut = mount(<Error {...mandatoryProps} />);

    expect(sut).toMatchSnapshot();
  });

  it('should render correctly with full props', () => {
    const props = {
      ...mandatoryProps,
      errorInfo: 'Info',
      showReload: true,
    };
    const sut = mount(<Error {...props} />);

    expect(sut).toMatchSnapshot();
  });

  it('should reload page after click', () => {
    const props = {
      ...mandatoryProps,
      errorInfo: 'Info',
      showReload: true,
    };
    const sut = mount(<Error {...props} />);

    expect(reloadSpy).toHaveBeenCalledTimes(0);

    sut.find(Button).simulate('click');

    expect(reloadSpy).toHaveBeenCalledTimes(1);
  });
});
