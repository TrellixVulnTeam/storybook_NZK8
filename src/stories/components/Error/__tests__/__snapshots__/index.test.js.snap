// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Error /> should render correctly with full props 1`] = `
<Error
  error="ERROR"
  errorInfo="Info"
  showReload={true}
  title="title"
>
  <_default
    withMargin={true}
  >
    <styled__Wrapper>
      <div
        className="styled__Wrapper-sc-1g40xjj-0 fFtIze"
      >
        <styled__InnerWrapper
          className="_test-inner-wrapper"
          withMargin={true}
        >
          <div
            className="_test-inner-wrapper styled__InnerWrapper-sc-1g40xjj-1 bRVfRH"
          >
            <styled__PageTitle>
              <h3
                className="styled__PageTitle-sc-160agpm-0 dnrbTB"
              >
                title
              </h3>
            </styled__PageTitle>
            <_default
              onClick={[Function]}
              pill={true}
              theme="blue-gradient"
            >
              <styled__Wrapper
                onClick={[Function]}
                pill={true}
                theme="blue-gradient"
                type="button"
              >
                <button
                  className="styled__Wrapper-sc-6m8pxr-0 bJTQMf"
                  onClick={[Function]}
                  type="button"
                >
                  Refresh
                </button>
              </styled__Wrapper>
            </_default>
            <styled__Details>
              <details
                className="styled__Details-sc-160agpm-1 cyDxZW"
              >
                <summary>
                  ERROR
                </summary>
                <br />
                Info
              </details>
            </styled__Details>
          </div>
        </styled__InnerWrapper>
      </div>
    </styled__Wrapper>
  </_default>
</Error>
`;

exports[`<Error /> should render correctly with mandatory props 1`] = `
<Error
  error="ERROR"
  title="title"
>
  <_default
    withMargin={true}
  >
    <styled__Wrapper>
      <div
        className="styled__Wrapper-sc-1g40xjj-0 fFtIze"
      >
        <styled__InnerWrapper
          className="_test-inner-wrapper"
          withMargin={true}
        >
          <div
            className="_test-inner-wrapper styled__InnerWrapper-sc-1g40xjj-1 bRVfRH"
          >
            <styled__PageTitle>
              <h3
                className="styled__PageTitle-sc-160agpm-0 dnrbTB"
              >
                title
              </h3>
            </styled__PageTitle>
            <styled__Details>
              <details
                className="styled__Details-sc-160agpm-1 cyDxZW"
              >
                <summary>
                  ERROR
                </summary>
                <br />
              </details>
            </styled__Details>
          </div>
        </styled__InnerWrapper>
      </div>
    </styled__Wrapper>
  </_default>
</Error>
`;
